File: src/R/src/constrained_boosting.r
library(assertthat)
library(collections)
library(MASS)
library(NlcOptim)
library(mltools)
library(data.table)

source("../src/data_structures.r")

cross_validate<- function(data,synth_effect,
                          n_trees,max_depth=3,regularization = 2,
                          min_samples_leaf=1,nfolds=5){
  assert_that(is(data,"data"))
  assert_that(is.numeric(synth_effect))
  assert_integer(n_trees)
  assert_integer(max_depth)
  assert_that(is.numeric(regularization))
  assert_integer(min_samples_leaf)
  assert_integer(nfolds)
  
  #training_error<-vector(mode = "list",length = nfolds)
  test_error<-vector(mode = "list",length = nfolds)
  
  n_fold_elements<-data$N*(1-1/nfolds)
  
  for(i in 1:nfolds){
    #There are two nfolds parameters in this function.The nfold of the
    #function below states that the data will be splitted into [nfolds] sets. 
    #The first of them has N/nfolds elements, the second - N*(nfolds-1)/nfolds
    #In this loop we do nfold(from my function) such splits and always take the
    #first fold for training and the second for test.
    
    fold <- folds(W,nfolds = data$N/n_fold_elements,stratified = TRUE)
    tr<-which(f==1)
    te<-which(f==2)
    
    F<-constrained_boost(data,synth_effect,
                         n_trees,max_depth,regularization,
                         min_samples_leaf,tr)
    
    #training_error[[i]]<-vector(mode ="numeric",length = (n_trees+1))
    test_error[[i]]<-vector(mode ="numeric",length = (n_trees+1))
    
    for(j in 1:(n_trees+1)){
      #training_error[[i]][[j]]<-sum(evaluate(data$Y[tr],F[[j]]$observed$`TRUE`[tr]))/length(tr)
      test_error[[i]][[j]]<-sum(evaluate(data$Y[te],F[[j]]$observed$`TRUE`[te]))/length(te)
    }
  }
  #return (list(training_error=training_error,test_error=test_error))
  return(test_error)
}

constrained_boost <- function(data,synth_effect,
                              n_trees,max_depth=3,regularization = 2,
                              min_samples_leaf=1,tr=TRUE){
  assert_that(is(data,"data"))
  assert_that(is.numeric(synth_effect))
  assert_integer(n_trees)
  assert_integer(max_depth)
  assert_that(is.numeric(regularization))
  assert_integer(min_samples_leaf)
  assert_that(is.logical(tr)||is.numeric(tr))
  
  const_pair = fit_const_pair(get_index(data,tr),synth_effect)
  
  F<-vector(mode="list",length = n_trees)
  F[[1]]<-predict_counterfactuals(const_pair,data)
  
  residuals<-obs_gradient(data$Y,data$W,F[[1]],ind = tr)

  for (i in 1:n_trees){
    tree_pair<-fit_tree_pair(get_index(data,tr),residuals,max_depth,min_samples_leaf)
    
    leaf_index_dict<-build_leaf_index(tree_pair,data$X)
    
    v<-step_search(data,F[[i]],leaf_index_dict,regularization)
    
    F[[i+1]]<-add_counterfactuals(F[[i]],predict_counterfactuals(tree_pair,data,v))
    
    residuals<-obs_gradient(data$Y,data$W,F[[i+1]],ind=tr)
  }
  return(F)
}

fit_tree_pair<-function(data,residuals,max_depth,min_samples_leaf){
  assert_that(is(data,"data"))
  assert_that(is(residuals,"numeric_treatment_dictionary"))
  assert_integer(max_depth)
  assert_integer(min_samples_leaf)
  
  tree_pair<-TreatmentDictionary(0,0)
  for(tr in names(tree_pair)){
    tr_logical<-as.logical(tr)
    tree<-fit_regression_tree(get_elements_by_treatment(data$X,data$W,t=tr_logical), 
                              residuals[[tr]],
                              min_samples_leaf=min_samples_leaf,
                              max_depth=max_depth)
    tree_pair[[tr]]<-tree
  }
  tree_pair
}

fit_const_pair <- function(data,synth_effect){
  assert_that(is(data,"data"))
  assert_that(is.numeric(synth_effect))
  
  mean <- mean(data$Y)
  dict<-TreatmentDictionary(0,0)
  for(tr in names(dict)){
    tr_logical<-as.logical(tr)
    value<-mean+(2*tr_logical-1)*synth_effect*data$N_treated[[tr]]/data$N
    dict[[tr]]<-Leaf(value)
  }
  dict
}

predict_counterfactuals <- function(estimator_pair, data,v=NULL){
  assert_that(is(estimator_pair,"treatment_dictionary"))
  assert_that(is(estimator_pair$`TRUE`,"leaf")||is(estimator_pair$`TRUE`,"node"))
  assert_that(is(data,"data"))
  assert_that(is.null(v)||is(v,"treatment_dictionary"))
  
  if(is.null(v)){
    return (Counterfactuals(predict_treated_dictionary(estimator_pair,data$X),data$W))
  }
  else{
    treated<-NumericTreatmentDictionary(vector(mode = "numeric",length = length(data$W)),
                                        vector(mode = "numeric",length = length(data$W)))
    for(tr in names(estimator_pair)){
      assignments<-predict_rt_matrix(estimator_pair[[tr]],data$X,TRUE)
      
      for(i in 1:length(assignments)){
        treated[[tr]][[i]]<-v[[tr]]$get(as.character(assignments[i]))
      }
    }
    return (Counterfactuals(treated,data$W))
  }
 
}


obs_gradient <- function(Y,W,F,ind=TRUE){
  assert_that(is.numeric(Y))
  assert_that(is.logical(W))
  assert_that(is(F,"counterfactuals"))
  assert_that(is.logical(ind)||is.numeric(ind))
  
  dict<-NumericTreatmentDictionary(0,0)
  for(tr in names(dict)){
    tr_logical<-as.logical(tr)

    value<- neg_grad(get_elements_by_treatment(Y[ind],W[ind],tr_logical),
                     get_elements_by_treatment(get_index(F,ind),W[ind],tr_logical,TRUE))
    dict[[tr]]<-value
  }
  dict
}

build_leaf_index<-function(tree_pair,X){
  assert_that(is(tree_pair,"treatment_dictionary"))
  assert_that(is.matrix(X)&&is.numeric(X))
  
  assignments<-predict_treated_dictionary(tree_pair,X,TRUE)
 
  assignment_index_treatment_dictionary <-TreatmentDictionary(Dict$new(),Dict$new())
  for(tr in names(assignments)){
    #Unique leaves' indeces for treatment
    leaves<-unique(assignments[[tr]])
    for(l in leaves){
      indeces<-which(assignments[[tr]]==l)
      assignment_index_treatment_dictionary[[tr]]$set(as.character(l),indeces)
    }
  }
  assignment_index_treatment_dictionary
}

step_search<-function(data, F, assignment_indeces_dict,regularization,ind=TRUE){
  assert_that(is(data,"data"))
  assert_that(is(F,"counterfactuals"))
  assert_that(is(assignment_indeces_dict,"treatment_dictionary"))
  assert_that(is.numeric(regularization))
  assert_that(is.logical(ind)||is.numeric(ind))
  
  F1 <- F$treated$`TRUE`
  F0 <- F$treated$`FALSE`
  Y <- data$Y
  W <- data$W
  
  idx1<-assignment_indeces_dict$`TRUE` 
  idx0<-assignment_indeces_dict$`FALSE`
  
  filter1<-which(W==1)
  filter0<-which(W==0)
  
  if(!is.logical(ind)){
    filter1<-intersect(filter1,ind)
    filter0<-intersect(filter0,ind)
  }
  
  idx1obs <- filter_index(idx1, filter1)
  idx0obs <-filter_index(idx0, filter0)
  
  n1<-length(idx1$keys())
  n0<-length(idx0$keys())
  
  v<-rep(0.1,times=n1+n0)
  
  objfun=function(x){
    fun<-0
    for (i in 1:length(idx1obs$keys())){
      k<-idx1obs$keys()[i]
      sum_loss_key<-0
      for(j in idx1obs$get(k)){
        sum_loss_key<-sum_loss_key+(Y[j] - F1[j] - x[i])^2
      }
      fun<- fun+sum_loss_key
    }
    for (i in 1:length(idx0obs$keys())){
      k<-idx0obs$keys()[i]
      sum_loss_key<-0
      for(j in idx0obs$get(k)){
        sum_loss_key<-sum_loss_key+(Y[j] - F0[j] - x[n1+i])^2
      }
      fun<- fun+sum_loss_key
    }
    for (i in 1:length(idx1$keys())){
      k<-idx1$keys()[i]
      sum_loss_key<-0
      for(j in idx1$get(k)){
        sum_loss_key<-sum_loss_key + regularization*x[i]^2
      }
      fun<- fun+sum_loss_key
    }
    for (i in 1:length(idx0$keys())){
      k<-idx0$keys()[i]
      sum_loss_key<-0
      for(j in idx0$get(k)){
        sum_loss_key<-sum_loss_key + regularization*x[n1+i]^2
      }
      fun<- fun+sum_loss_key
    }
    
    return(fun)
  }
  
  confun=function(x){
    fun1<-0
    for (i in 1:length(idx1$keys())){
      k<-idx1$keys()[i]
      sum_loss_key<-0
      for(j in idx1$get(k)){
        sum_loss_key<-sum_loss_key + x[i]
      }
      fun1<- fun1+sum_loss_key
    }
    fun0<-0
    for (i in 1:length(idx0$keys())){
      k<-idx0$keys()[i]
      sum_loss_key<-0
      for(j in idx0$get(k)){
        sum_loss_key<-sum_loss_key +x[n1+i]
      }
      fun0<- fun0+sum_loss_key
    }
    
    f=NULL
    f=rbind(f,fun1-fun0)
    return(list(ceq=f,c=NULL))
  }
  
  v<-solnl(v,objfun=objfun,confun=confun)$par[,1]
  vopt1<-v[1:n1]
  vopt0<-v[(n1+1):(n1+n0)]
  
  dict<-TreatmentDictionary(0,0)
  for(tr in names(dict)){
    dict[[tr]]<-Dict$new()
    if(as.logical(tr)){
      for (i in 1:length(idx1$keys())){
        k<-idx1$keys()[i]
        dict[[tr]]$set(k,vopt1[i])
      }
    }
    else{
      for (i in 1:length(idx0$keys())){
        k<-idx0$keys()[i]
        dict[[tr]]$set(k,vopt0[i])
      }
    }
  }
  dict
}

filter_index<-function(indices, include_subjects){
  filtered_index = Dict$new()
  for (k in indices$keys()){
    v_in <- intersect(indices$get(k),include_subjects)
    filtered_index$set(k,v_in) 
  }
  filtered_index
}

File: src/Julia/constrained_boosting.jl
# note that when using the binomial loss, the treatment effect must be between -1 and 1
# TO DO:
# - Get rid of learning rate mechanic- prove that it works in the continuous case but moves us off the constraint surface with binary
# - Initialize the step seach leaf values with those computed by the trees
# - Add in a regularizer to do the job of the learning rate

__precompile__()

module CausalEffectMorph

export ObsData,
       Squared, Binomial,
       constrained_boost, cross_validate

include_dependency("data_structures.jl")
include_dependency("losses.jl")
include_dependency("estimators.jl")

include("data_structures.jl")
include("losses.jl")
include("estimators.jl")

using .Estimators
using .Losses
using .ObsDataStructures

using DataStructures
using JuMP, Ipopt

const TF = [true, false]

function fit_const_pair(data::ObsData, loss::Squared, τ::Real)
    Ȳ = mean(data.Y.Y)
    return Dict(t => Leaf(Ȳ +(2*t-1) * τ * data.N_treated[t]/data.N) for t in TF)
end

function fit_const_pair(data::ObsData, loss::Binomial, τ::Real)
    m = Model(solver=IpoptSolver(print_level=0))
    @variable(m, 0 <= πt <= 1)
    @variable(m, 0 <= πf <= 1)
    @constraint(m, πt-πf == τ)
    @NLobjective(m, Max, data.N_treated[true]*log(1-πf) +
                         data.N_treated[false]*log(πt))
    status = solve(m)
    π = Dict(true=>getvalue(πt), false=>getvalue(πf))
    return Dict(t  => Leaf(log(π[t]/(1-π[t]))) for t in TF)
end

function predict_counterfactuals{T<:Estimators.Estimator}(estimator_pair::Dict{Bool,T}, data::ObsData)
    return Counterfactuals(predict(estimator_pair, data.X.X), data.W)
end

function obs_gradient{T<:Loss}(loss::T, Y::Outcome, F::Counterfactuals; idx=:)
    return Dict(t => neg_grad(loss, Γ(Y[idx],t=t), Γ(F[idx],o=true,t=t)) for t in TF)
end

function build_leaf_index{T<:Estimators.Estimator}(tree_pair::Dict{Bool,T}, X::Covariates)
    assignments = Dict(t=>assign_leaves(tree_pair[t], X.X) for t in TF)
    aa = vcat([ls for (t,ls) in assignments]...)
    leaves = Set(vcat([ls for (t,ls) in assignments]...))
    assignment_index_matrix = [(t,l,find(assignments[t].==l)) for l in leaves, t in TF]
    assignment_index = OrderedDict((t,l)=>ind for (t,l,ind) in vec(assignment_index_matrix) if length(ind)>0)
    return assignment_index
end

function filter_index(index, include_subjects)
    filtered_index = OrderedDict()
    for (k,v) in index
        v_in = v ∩ include_subjects
        length(v_in)>0 ? filtered_index[k] = v_in : nothing
    end
    return filtered_index
end


function step_search(loss::Squared, data::ObsData, F::Counterfactuals, idx::OrderedDict, λ, τ=nothing; tr=:)
    F1 = F.treated[true]
    F0 = F.treated[false]
    Y = data.Y.Y
    W = data.W

    idx1, idx0 = OrderedDict(), OrderedDict()
    for ((t,l),i) in idx
        t ? idx1[l] = i : idx0[l] = i
    end

    idx1obs = filter_index(idx1, findin(W,1) ∩ tr)
    idx0obs = filter_index(idx0, findin(W,0) ∩ tr)

    m = Model(solver=IpoptSolver(print_level=0))
    @variable(m, ν1[1:length(idx1)])
    @variable(m, ν0[1:length(idx0)])
    @constraint(m, sum(sum(ν1[i] for j in idx1[l]) for (i,l) in enumerate(keys(idx1))) -
                   sum(sum(ν0[i] for j in idx0[l]) for (i,l) in enumerate(keys(idx0))) == 0)
    @objective(m, Min,
                   sum(sum((Y[j] - F1[j] - ν1[i])^2 for j in idx1obs[l]) for (i,l) in enumerate(keys(idx1obs))) +
                   sum(sum((Y[j] - F0[j] - ν0[i])^2 for j in idx0obs[l]) for (i,l) in enumerate(keys(idx0obs))) +
                   sum(sum(λ*ν1[i]^2 for j in idx1[l]) for (i,l) in enumerate(keys(idx1))) +
                   sum(sum(λ*ν0[i]^2 for j in idx0[l]) for (i,l) in enumerate(keys(idx0))))
    status = solve(m)
    νopt1 = getvalue(ν1)
    νopt0 = getvalue(ν0) # this is fucked. even when I add the learning rate back it's still broken...
    return Dict(true => Dict(l=>νi for (l,νi) in zip(keys(idx1),νopt1)),
                false => Dict(l=>νi for (l,νi) in zip(keys(idx0),νopt0)))
end

function step_search(loss::Binomial, data::ObsData, F::Counterfactuals, idx::OrderedDict, λ, τ; tr=:)
    F1 = F.treated[true]
    F0 = F.treated[false]
    Y = data.Y.Y
    W = data.W

    idx1, idx0 = OrderedDict(), OrderedDict()
    for ((t,l),i) in idx
        t ? idx1[l] = i : idx0[l] = i
    end

    idx1obs = filter_index(idx1, findin(W,1) ∩ tr)
    idx0obs = filter_index(idx0, findin(W,0) ∩ tr)

    m = Model(solver=IpoptSolver(print_level=0))
    @variable(m, ν1[1:length(idx1)])
    @variable(m, ν0[1:length(idx0)])
    @NLconstraint(m, sum(sum((1+exp(-F1[j]-ν1[i]))^(-1) for j in idx1[l]) for (i,l) in enumerate(keys(idx1))) -
                     sum(sum((1+exp(-F0[j]-ν0[i]))^(-1) for j in idx0[l]) for (i,l) in enumerate(keys(idx0))) == data.N*τ)
    @NLobjective(m, Min,
                    sum(sum(log(1+exp(F1[j]+ν1[i])) - Y[j]*(F1[j]+ν1[i]) for j in idx1obs[l]) for (i,l) in enumerate(keys(idx1obs))) +
                    sum(sum(log(1+exp(F0[j]+ν0[i])) - Y[j]*(F0[j]+ν0[i]) for j in idx0obs[l]) for (i,l) in enumerate(keys(idx0obs))) +
                    sum(sum(λ*ν1[i]^2 for j in idx1[l]) for (i,l) in enumerate(keys(idx1))) +
                    sum(sum(λ*ν0[i]^2 for j in idx0[l]) for (i,l) in enumerate(keys(idx0))))
    status = solve(m)
    νopt1 = getvalue(ν1)
    νopt0 = getvalue(ν0)
    return Dict(true => Dict(l=>νi for (l,νi) in zip(keys(idx1),νopt1)),
                false => Dict(l=>νi for (l,νi) in zip(keys(idx0),νopt0)))
end

function predict_counterfactuals(tree_pair, X::Covariates, ν)
    return Counterfactuals(Dict(t=>[ν[t][li] for li in assign_leaves(tree_pair[t], X.X)] for t in TF), X.W)
end

function constrained_boost{T<:Real, T2<:Real, L<:Loss}(data::ObsData, loss::L, τ::T, n_trees::Int;
                           max_depth::Int=3, λ::T2=2, min_samples_leaf::Int=1,
                           tr=:)
    F = Vector{Counterfactuals}(0)

    const_pair = fit_const_pair(data[tr], loss, τ)
    predicted_counterfactuals = predict_counterfactuals(const_pair, data)
    push!(F, predicted_counterfactuals)
    residuals = obs_gradient(loss, data.Y, F[end], idx=tr)

    for i in 1:n_trees
        tree_pair = Dict(t=>fit_regression_tree(Γ(data.X[tr],t=t) , residuals[t],
                                                        min_samples_leaf=min_samples_leaf,
                                                        max_depth=max_depth) for t in TF);
        idx = build_leaf_index(tree_pair, data.X)
        ν = step_search(loss, data, F[end], idx, λ, τ, tr=tr)
        push!(F, F[end]+predict_counterfactuals(tree_pair, data.X, ν))
        residuals = obs_gradient(loss, data.Y, F[end], idx=tr)
    end
    return F
end

import MLBase.StratifiedKfold

function cross_validate{T<:Real, T2<:Real, L<:Squared}(data::ObsData, loss::L, τ::T, n_trees::Int;
                           max_depth::Int=3, λ::T2=2, min_samples_leaf::Int=1,
                           nfolds::Int=5)
    training_error, test_error = Vector{Vector{Float64}}(0), Vector{Vector{Float64}}(0)
    loss == Binomial() ? folds = StratifiedKfold(data.Y.Y + 2*data.W, nfolds) : folds = StratifiedKfold(data.W, nfolds)
    for tr in folds
        println("tr")
        println(tr)
        te = [i for i in 1:data.N if i ∉ tr]
        F = constrained_boost(data, loss, τ, n_trees,
                           max_depth=max_depth, λ=λ, min_samples_leaf=min_samples_leaf,
                           tr=tr)
        push!(training_error, [sum(evaluate(loss, data.Y.Y[tr], F[n].observed[true][tr]))/length(data.Y.Y[tr]) for n in 1:n_trees+1])
        push!(test_error, [sum(evaluate(loss, data.Y.Y[te], F[n].observed[true][te]))/length(data.Y.Y[te]) for n in 1:n_trees+1])
    end
    return training_error, test_error
end

end # module
